### ISSUE TITLE
Groups and apps can apparently overlap and it creates weird behaviour

### ISSUE THREAD

**Issue by pierrecdn** [mesosphere/marathon#4901]
Hi,
In 1.3 I have apps that were deployed using `/v2/apps`and `/v2/groups` (probably using `?force=true`at some point). The issue is that now they are visible at 2 levels when querying `/v2/groups`

```
$ curl -kL -u ${MARATHON_USER}:${MARATHON_PASSWORD} https://my-marathon.mytld/v2/groups \
    | jq '[.groups[].apps[].id], [.apps[].id]'
[
  (...)
  "/group/app1",
  "/group/app2",
  (...)
],
[
  (...)
  "/group/app1",
  "/group/app2",
  (...)
]
```

The issue it produces: now I can't deploy anymore, all my new deployments are apparently "attached" to a deployment_id in which there are app1 and app2 in "Restarting" state (probably an action initiated by the owner). These 2 apps are themselves marked as "deploying".

![bug](https://cloud.githubusercontent.com/assets/4225738/21401345/4a6f7062-c7b4-11e6-9fbc-309c2e579ce2.png)

**Comment by pierrecdn** [comment]
So the issue apparently occurs when using `/v2/groups` with group "/" and appIds containing slashes ('/').
Then use `/v2/apps` with the same id(s) (add `?force=true` if needed), and the case should be reproducible.

@FlorentFlament can maybe gives more details.

**Comment by FlorentFlament** [comment]
Here are the steps to reproduce the issue. On an empty instance of Marathon listening on localhost:8080

```
# Create the 'hello' app in group '/test'
curl -H "Content-Type: application/json" -X POST -d '{"id":"/test/hello", "cmd":"while true; do echo hello; sleep 1; done"}' http://localhost:8080/v2/apps

# Then create the 'test/hello' app in group '/'
curl -H "Content-Type: application/json" -X PUT -d '{"id": "/", "apps": [{"id":"/test/hello", "cmd":"while true; do echo hello; sleep 1; done"}]}' http://localhost:8080/v2/groups

# Now Marathon sees 2 'running' applications with the same path: '/test/hello'
# Let's restart the '/test/hello' application
curl -X POST http://localhost:8080/v2/apps/test/hello/restart

# Now deploy the 'hello' app in group '/'
curl -H "Content-Type: application/json" -X POST -d '{"id":"/hello", "cmd":"while true; do echo hello; sleep 1; done"}' http://localhost:8080/v2/apps
```

As you can see, any deployment will now result in a 'stuck' deployment affecting both the legit app being deployed as well as the '/test/hello' app.

**Comment by FlorentFlament** [comment]
And here are the steps to work around the issue, by removing the bad 'test/hello' application from group '/':

```
# Backup configuration of applications in group '/'
curl http://localhost:8080/v2/groups | jq .apps > marathon-backup.json

# Remove bad entries from the file (i.e all applications with an id including multiple '/')
cp marathon-backup.json marathon-cleaned.json  # Edit marathon-cleaned.json

# Redeploy all applications of group '/'
curl -X PUT -d "{\"id\": \"/\", \"apps\": $(cat marathon-cleaned.json)}" http://localhost:8080/v2/groups?force=true
```

No that this is potentially dangerous, since it 'reconfigures' all the applications of the '/' group.

**Comment by FlorentFlament** [comment]
IMHO, Marathon should return an error message if the applications deployed through the /v2/groups endpoint are not in the group specified in the JSON document (i.e there is a discrepency between an app id and its group id).

**Comment by david-crowder** [comment]
Just ran into this as well. Posted to the `/` group with a definition that contained other groups and saw the behavior already described. Things got in a really weird/annoying state.

**Comment by aquamatthias** [comment]
To reproduce the problem easily, you can do this:
```curl -H "Content-Type: application/json" -X PUT -d '{"id": "/", "apps": [{"id":"/test/hello", "cmd":"while true; do echo hello; sleep 1; done"}]}' http://localhost:8080/v2/groups
```

This will create an app `/test/hello` in group `/` - which is wrong.

**Comment by aquamatthias** [comment]
Fix for master is here: https://phabricator.mesosphere.com/D518

**Comment by aquamatthias** [comment]
Fix for 1.4 + migration to repair the group hierarchy: https://github.com/mesosphere/marathon/pull/5179
